### PR TITLE
[RLlib; Offline RL] Implement twin-Q net option for CQL.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

This PR proposes the double Q trick for CQL to stabilize training. More specifically
- it modifies the loss computation to also include loss terms for the twin Q-value function
- it stores these additional loss terms to run a backward pass with optimizer for the twin Q-value networks

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/cql/torch/cql_torch_learner.py [hunk 1]
```diff
@@ -90,8 +92,9 @@ def compute_loss_for_module(
                 # Use the actions sampled from the current policy.
                 Columns.ACTIONS: actions_curr,
             }
+            # Note, if `twin_q` is `True`, `compute_q_values` computes the minimum
```
Comment (sven1977) [inline rllib/algorithms/cql/torch/cql_torch_learner.py:95]: Nice!

File: rllib/algorithms/cql/torch/cql_torch_learner.py [hunk 2]
```diff
@@ -144,15 +150,24 @@ def compute_loss_for_module(
         # Calculate the TD error.
         td_error = torch.abs(q_selected - q_selected_target)
         # TODO (simon): Add the Twin TD error
+        if config.twin_q:
+            td_error += torch.abs(q_twin_selected, q_selected_target)
+            # Rescale the TD error
+            td_error += 0.5
```
Comment (sven1977) [inline rllib/algorithms/cql/torch/cql_torch_learner.py:156]: Should this be `* 0.5`?

File: rllib/algorithms/cql/torch/cql_torch_learner.py [hunk 3]
```diff
@@ -144,15 +150,24 @@ def compute_loss_for_module(
         # Calculate the TD error.
         td_error = torch.abs(q_selected - q_selected_target)
         # TODO (simon): Add the Twin TD error
+        if config.twin_q:
+            td_error += torch.abs(q_twin_selected, q_selected_target)
```
Comment (sven1977) [inline rllib/algorithms/cql/torch/cql_torch_learner.py:154]: Should this be `torch.minimum`?

File: rllib/algorithms/cql/torch/cql_torch_learner.py [hunk 4]
```diff
@@ -144,15 +150,24 @@ def compute_loss_for_module(
         # Calculate the TD error.
         td_error = torch.abs(q_selected - q_selected_target)
         # TODO (simon): Add the Twin TD error
```
Comment (sven1977) [inline rllib/algorithms/cql/torch/cql_torch_learner.py:152]: Remove this TODO.

File: rllib/algorithms/cql/torch/cql_torch_learner.py (hunk 3 above)
Comment (sven1977) [reply rllib/algorithms/cql/torch/cql_torch_learner.py:154]: Sorry, meant: Should this be `torch.abs(q_twin_selected - q_selected_target)`

File: rllib/algorithms/cql/torch/cql_torch_learner.py [hunk 5]
```diff
@@ -231,23 +290,42 @@ def compute_loss_for_module(
...
+                    q_twin_curr_repeat - logps_curr_repeat.detach(),
+                ],
+                dim=1,
+            )
+            cql_twin_loss = (
+                torch.logsumexp(q_twin_repeat / config.temperature, dim=1).mean()
+                * config.min_q_weight
+                * config.temperature
+            )
+            cql_twin_loss - (q_twin_selected.mean()) * config.min_q_weight
```
Comment (sven1977) [inline rllib/algorithms/cql/torch/cql_torch_learner.py:317]: Should this be `-=`?

Could you check all the math also once more? I don't want to miss anything important here :)

File: rllib/algorithms/cql/torch/cql_torch_learner.py (hunk 2 above)
Comment (simonsays1980) [reply rllib/algorithms/cql/torch/cql_torch_learner.py:156]: Great catch :D This should be multiplied by 0.5

File: rllib/algorithms/cql/torch/cql_torch_learner.py (hunk 3 above)
Comment (simonsays1980) [reply rllib/algorithms/cql/torch/cql_torch_learner.py:154]: No, this is fine like this. We only add the TD-error. The Q-values are computed by using the minimum of Q- and twin-Q-value. In case of the TD error we would actually cancel out large errors, if we would do so. But I am open to discussion.

File: rllib/algorithms/cql/torch/cql_torch_learner.py (hunk 5 above)
Comment (simonsays1980) [reply rllib/algorithms/cql/torch/cql_torch_learner.py:317]: Oh yes good catch. I hope I just missed the `=` key and its not my eyes :)

**Review by sven1977 — APPROVED** [review]
LGTM now. Thanks for this great PR and for double-checking @simonsays1980 .
